pkg/restore/restore.go

```diff
@@ -578,6 +590,102 @@ func addToResult(r *api.RestoreResult, ns string, e error) {
+			pvLogger.Debug("PV not found, safe to restore")
+			// PV not found, can safely exit loop and proceed with restore.
+			shouldRestore = true
+			return true, nil
+		}
+
+		if err != nil {
+			return true, errors.Wrapf(err, "could not retrieve in-cluster copy of PV %s", name)
+
+		}
+
+		var namespace, pvcName string
+		if clusterPV != nil {
+			phase, err := collections.GetString(clusterPV.UnstructuredContent(), "status.phase")
+			if err != nil {
+				// Break the loop since we couldn't read the phase
+				return true, errors.Wrapf(err, "error getting phase for in-cluster PV %s", name)
+			}
+
+			if phase == string(v1.VolumeReleased) || clusterPV.GetDeletionTimestamp() != nil {
+				// PV was found and marked for deletion, or it was released; wait for it to go away.
+				pvLogger.Debugf("PV found, but marked for deletion, waiting")
+				return false, nil
+			}
+
+			namespace, err = collections.GetString(clusterPV.UnstructuredContent(), "spec.claimRef.namespace")
+			if err != nil {
+				return true, errors.Wrapf(err, "error looking up namespace name for in-cluster PV %s", name)
+			}
+			pvcName, err = collections.GetString(clusterPV.UnstructuredContent(), "spec.claimRef.name")
+			if err != nil {
+				return true, errors.Wrapf(err, "error looking up persistentvolumeclaim for in-cluster PV %s", name)
+			}
+		}
+
+		// Have to create the PVC client here because we don't know what namespace we're using til we get to this point.
+		pvcClient := ctx.coreClient.PersistentVolumeClaims(namespace)
+
+		pvc, err := pvcClient.Get(pvcName, metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			pvLogger.Debugf("PVC %s for PV not found, waiting", pvcName)
+			// PVC wasn't found, but the PV still exists, so continue to wait.
+			return false, nil
+		}
+
+		if err != nil {
+			return true, errors.Wrapf(err, "error getting claim %s for persistent volume", pvcName, name)
+		}
+
+		if pvc != nil && pvc.GetDeletionTimestamp() != nil {
+			pvLogger.Debugf("PVC for PV marked for deletion, waiting")
+			// PVC is still deleting, continue to wait.
+			return false, nil
+		}
+
+		ns, err := ctx.namespaceClient.Get(namespace, metav1.GetOptions{})
+		if apierrors.IsNotFound(err) {
+			pvLogger.Debugf("namespace %s for PV not found, waiting", namespace)
+			// namespace not found but the PV still exists, so continue to wait
+			return true, nil
+		}
+		if err != nil {
+			return true, errors.Wrapf(err, "error getting namespace %s associated with PV %s", namespace, name)
+		}
+
+		if ns != nil && (ns.GetDeletionTimestamp() != nil || ns.Status.Phase == v1.NamespaceTerminating) {
+			pvLogger.Debugf("namespace %s associated with PV is deleting, waiting", namespace)
+			// namespace is in the process of deleting, keep looping
+			return false, nil
+		}
+
+		// None of the PV, PVC, or NS are marked for deletion, break the loop.
+		pvLogger.Debug("PV, associated PVC and namespace are not marked for deletion")
+		return true, nil
+	})
+
+	end := time.Now()
```

Comment:
Check the `err` that `wait.PollImmediate` returns instead of comparing start/end times.

Reply:
I think I can just return the `err` that `wait.PollImmediate` uses, unless I want to add a debug line for timeout. It will be added to the result at L810 in this version of the PR.